crates/miden-objects/src/block/account_tree.rs

```diff
@@ -0,0 +1,458 @@
+            };
+
+            // SAFETY: Since we only inserted account IDs into the SMT, it is guaranteed that
+            // the leaf_idx is a valid Felt as well as a valid account ID prefix.
+            AccountTreeError::DuplicateStateCommitments {
+                prefix: AccountIdPrefix::new_unchecked(
+                    Felt::try_from(leaf_idx).expect("leaf index should be a valid felt"),
+                ),
+            }
+        })?;
+
+        for (leaf_idx, leaf) in smt.leaves() {
+            if leaf.num_entries() >= 2 {
+                // SAFETY: Since we only inserted account IDs into the SMT, it is guaranteed that
+                // the leaf_idx is a valid Felt as well as a valid account ID prefix.
+                return Err(AccountTreeError::DuplicateIdPrefix {
+                    duplicate_prefix: AccountIdPrefix::new_unchecked(
+                        Felt::try_from(leaf_idx.value())
+                            .expect("leaf index should be a valid felt"),
+                    ),
+                });
+            }
+        }
+
+        Ok(AccountTree { smt })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an opening of the leaf associated with the `account_id`. This is a proof of the
+    /// current state commitment of the given account ID.
+    ///
+    /// Conceptually, an opening is a Merkle path to the leaf, as well as the leaf itself.
+    pub fn open(&self, account_id: AccountId) -> AccountWitness {
+        let key = Self::id_to_smt_key(account_id);
+        let proof = self.smt.open(&key);
+
+        // Check which account ID this proof actually contains. We rely on the fact that the tree
+        // only contains zero or one entry per account ID prefix.
+        //
+        // If the requested account ID matches an existing ID's prefix but their suffixes do not
+        // match, then this witness is for the _existing ID_.
+        //
+        // Otherwise, if the ID matches the one in the leaf or if it's empty, the witness is for the
+        // requested ID.
+        let witness_id = match proof.leaf() {
+            SmtLeaf::Empty(_) => account_id,
+            SmtLeaf::Single((key_in_leaf, _)) => {
+                // SAFETY: By construction, the tree only contains valid IDs.
+                Self::smt_key_to_id(*key_in_leaf)
+            },
+            SmtLeaf::Multiple(_) => {
+                unreachable!("account tree should only contain zero or one entry per ID prefix")
+            },
+        };
+
+        let commitment = Digest::from(
+            proof
+                .get(&Self::id_to_smt_key(witness_id))
+                .expect("we should have received a proof for the witness key"),
+        );
+
+        // SAFETY: The proof is guaranteed to have depth AccountTree::DEPTH.
+        AccountWitness::new_unchecked(witness_id, commitment, proof.into_parts().0)
```

Comment:
Same as the previous comment: could this be moved into something like `AccountWitness::from_smt_proof()`?